feat: a series of smooth functions that converges (locally) uniformly is smooth

CoolRmal opened the pull request:
The main theorem is the proof that an infinite sum of $C^n$ functions on a one-dimensional domain which converges locally uniformly is $C^n$ and related results.

This should be compared with [contDiff_tsum](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/SmoothSeries.html#contDiff_tsum) where the same result is proved for functions on an arbitrary domain but with a slightly stronger convergence assumption.

The main motivation for this PR is the [#mathlib4 > Tychonov's Counterexample for the Heat Equation](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Tychonov's.20Counterexample.20for.20the.20Heat.20Equation/with/547367912).

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

CoolRmal committed:
Initial commit (orphan branch)

CoolRmal committed:
Merge branch 'leanprover-community:master' into unique_riesz

CoolRmal committed:
Merge branch 'leanprover-community:master' into main

CoolRmal committed:
Merge branch 'leanprover-community:master' into main

CoolRmal committed:
Merge branch 'leanprover-community:master' into main

CoolRmal committed:
Merge branch 'leanprover-community:master' into main

CoolRmal committed:
Merge branch 'leanprover-community:master' into main

CoolRmal committed:
SummableLocallyUniformlyOn.contDiffOn_tsum

CoolRmal committed:
Merge branch 'leanprover-community:master' into contDiffwithinTsum

CoolRmal committed:
change name

CoolRmal committed:
Merge branch 'contDiffwithinTsum' of https://github.com/CoolRmal/mathlib4 into contDiffwithinTsum

ocfnash commented on Mathlib/Topology/Algebra/InfiniteSum/UniformOn.lean:
```diff
@@ -117,6 +117,16 @@ theorem MultipliableUniformlyOn.hasProdUniformlyOn [T2Space α] (h : Multipliabl
   exact (hasProdUniformlyOn_iff_tendstoUniformlyOn.mp hg s hs).congr_right (hg.tprod_eqOn hs).symm
 
+@[to_additive]
+lemma MultipliableUniformlyOn_congr [T2Space α]
```
```suggestion
lemma MultipliableUniformlyOn.congr [T2Space α]
```
(and then you can use dot notation later on)

CoolRmal committed:
Update Mathlib/Topology/Algebra/InfiniteSum/UniformOn.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

CoolRmal committed:
Update Mathlib/Topology/Algebra/InfiniteSum/UniformOn.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

CoolRmal committed:
Update TsumUniformlyOn.lean

CoolRmal commented on Mathlib/Topology/Algebra/InfiniteSum/TsumUniformlyOn.lean:
```diff
@@ -58,20 +56,75 @@ lemma SummableLocallyUniformlyOn_of_locally_bounded [TopologicalSpace β] [Local
+  have := hasSumUniformlyOn_iff_tendstoUniformlyOn.mp h.hasSumUniformlyOn s (Set.mem_singleton s)
+  refine this.continuousOn ?_
+  filter_upwards with _
+  fun_prop
```
Yes this is a generalization, but [continuousOn_tsum](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Normed/Group/FunctionSeries.html#continuousOn_tsum) is probably easier to use (at least I believe it is a more common of proving that the infinite sum of continuous functions is continuous).

CoolRmal committed:
Merge branch 'leanprover-community:master' into contDiffwithinTsum

CoolRmal committed:
fix

CoolRmal committed:
Update UniformOn.lean

ocfnash commented on Mathlib/Topology/Algebra/InfiniteSum/UniformOn.lean:
```diff
@@ -128,6 +128,14 @@ lemma MultipliableUniformlyOn.mono {t : Set β}
+    (h2 : MultipliableUniformlyOn f t) : MultipliableUniformlyOn f' t := by
+  refine (h2.hasProdUniformlyOn.congr ?_).multipliableUniformlyOn
+  filter_upwards with n
+  exact eqOn_fun_finsetProd h n
```
Style nit:
```suggestion
    (h : ∀ (i : ι), Set.EqOn (f i) (f' i) t) (h2 : MultipliableUniformlyOn f t) :
    MultipliableUniformlyOn f' t :=
  (h2.hasProdUniformlyOn.congr (.of_forall <| eqOn_fun_finsetProd h)).multipliableUniformlyOn
```

ocfnash reviewed (commented):
I'm afraid I only have time for a very superficial review but I'll try to return quickly after you've had a chance to respond.

Please also merge master since you are changing imports and I'd like to see the results with https://github.com/leanprover-community/mathlib4/pull/32133 applied.

CoolRmal committed:
Merge branch 'master' into contDiffwithinTsum

CoolRmal committed:
Update Mathlib/Topology/Algebra/InfiniteSum/TsumUniformlyOn.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

CoolRmal committed:
Update Mathlib/Topology/Algebra/InfiniteSum/TsumUniformlyOn.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

CoolRmal committed:
Update Mathlib/Topology/Algebra/InfiniteSum/UniformOn.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

CoolRmal committed:
Update Mathlib/Topology/Algebra/InfiniteSum/TsumUniformlyOn.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

CoolRmal committed:
Update Mathlib/Topology/Algebra/InfiniteSum/TsumUniformlyOn.lean

Co-authored-by: Oliver Nash <7734364+ocfnash@users.noreply.github.com>

ocfnash committed:
Fix trivial breakages

CoolRmal committed:
update

CoolRmal commented:
I deleted theorems starting with `SummableUniformlyOn`. I also deleted theorems starting with `SummableUniformly` because we have [summableUniformlyOn_univ_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/InfiniteSum/UniformOn.html#summableUniformlyOn_univ_iff).

CoolRmal committed:
Update TsumUniformlyOn.lean

ocfnash committed:
Final tweaks

Specifically:
 * Add a cross-reference to `contDiff_tsum` doc string
 * Tidy up sectioning
 * Add missing deprecation aliases for namespacing
 * Move and `to_additive`ise `SummableLocallyUniformlyOn.continuousOn_tsum`

ocfnash commented:
@CoolRmal I was finally able to take a close look at this today. I hope you don't mind but I decided to push up [a small commit](https://github.com/leanprover-community/mathlib4/pull/31377/commits/8d5a2c3b290151e910a8840ae98ac1921d44925a) since some of the suggestions were a bit tricky to explain through the GitHub interface.

I now understand that what this PR is doing is proving [https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/SmoothSeries.html#contDiff_tsum](contDiff_tsum) with a slightly weaker assumption but only for one-dimensional domains. I would have hoped that we could have a single result which dominated both of these lemmas. However I note that the file `TSumUniformlyOn.lean` has to some extent already established this precedent. With that in mind I'm going to ask another maintainer to give an opinion here.

ocfnash commented:
> I'd rather have directly the higher-dimensional result, i.e., a version of `contDiff_tsum` requiring only local uniform convergence. Is that really harder to prove?

Thanks, I agree. Let's aim for that instead! Would you like to take this on @CoolRmal ?

CoolRmal committed:
Merge branch 'master' into contDiffwithinTsum
